### DISCUSSION TITLE
How does this library interprets the sign in the DBC format?

### DISCUSSION THREAD

**Discussion by gpollo** [cantools/cantools#325] [General]
In the DBC format, a little-endian unsigned signal is defined with `@1+`. Likewise, a little-endian signed signal is defined with `@1-`. What I'm trying to understand, is which sign is it referring to.

The way I understand it, this sign is how the **raw value** should be interpreted, not the sign of the **physical value** since its sign can easily be inferred from the min/max. Signed raw signals would be encoded using two's complement while unsigned signals would encoded with no sign bit.

For example, let's take this signal: `@1+ (1,0) [-40|40]`. It is a little-endian unsigned signal. The scale/offset are 1/0. The min/max are -40/40. I should be able to send a physical value of -10, right? I usually get the following exception if I try doing that with the library:
```
OverflowError: can't convert negative int to unsigned
```
So, I believe that the library interprets the sign in a DBC signal as the sign of the **physical value**, hence it can't pack it. Now, I'm not saying it is wrong, I just find it odd. Why would the DBC format bother to have a sign in the signal if it could be inferred from min/max alone? Should this library throw an error while parsing the DBC if the min/max is not supported by the sign of the physical value?

There is also this [page](http://socialledge.com/sjsu/index.php/DBC_Format) which says that you should always defined an unsigned signal even though it is fair to assume that sending negative physical values is an important use case. I understand that the DBC format is proprietary and I can't find any official information about it. Does anybody has any more information about this? Documents maybe? Thanks!

**Comment by juleq** [comment]
Yes, endianness and signedness always refer to the packed (raw) data. The signedness tells the decoder at the other end if twos-complement has to be considered for any received int type or not.

On reception, the packed bits are read into a suitable host data type respecting endianness and signedness information. Following that the receiver applies factor and offset to translate from raw to physical.

If the receiver chooses to check validity of the received physical value, he can do so by testing if it is inside the specified physical signal range.

I would not expect surprises in cantools on that end, but if you provide your full signal spec (the SG text line) and your calls to cantools that fail I could probably comment.

**Reply by juleq** [reply]
Your mistake is probably having the signal specified as unsigned. Replace that + with a - and you should be good.

**Reply by juleq** [reply]
To answer "I should be able to send a physical value of -10, right?" 
Wrong. You can only encode a negative physical value by using signed OR unsigned and an offset to shift the physical value into negative. Your signal spec was unsigned and had an offset of 0. That results in the smallest representable physical value of 0.
